segmentation/utils/spatial_analysis.py

```diff
@@ -0,0 +1,245 @@
+    return close_num_rand
+
+
+def calculate_enrichment_stats(close_num, close_num_rand):
+    """Calculates z score and p values from spatial enrichment analysis.
+
+    Args:
+        close_num: marker x marker matrix with counts for cells
+            positive for corresponding markers
+        close_num_rand: random positive marker counts
+            for every permutation in the bootstrap
+
+    Returns:
+        z: z scores for corresponding markers
+        muhat: predicted mean values of close_num_rand random distribution
+        sigmahat: predicted standard deviation values of close_num_rand
+            random distribution
+        p: p values for corresponding markers, for both positive
+            and negative enrichment
+        h: matrix indicating whether
+            corresponding marker interactions are significant
+        adj_p: fdh_br adjusted p values"""
+    # Get the number of markers and number of permutations
+    marker_num = close_num.shape[0]
+    bootstrap_num = close_num_rand.shape[2]
+
+    # Create z, muhat, sigmahat, and p
+    z = np.zeros((marker_num, marker_num))
+    muhat = np.zeros((marker_num, marker_num))
+    sigmahat = np.zeros((marker_num, marker_num))
+    p = np.zeros((marker_num, marker_num, 2))
+
+    for j in range(0, marker_num):
+        for k in range(0, marker_num):
+            # Get close_num_rand value for every marker combination and reshape to use as input for norm fit
+            tmp = np.reshape(close_num_rand[j, k, :], (bootstrap_num, 1))
+            # Get muhat and sigmahat values for distribution from 100 permutations
+            (muhat[j, k], sigmahat[j, k]) = scipy.stats.norm.fit(tmp)
+            # Calculate z score based on distribution
+            z[j, k] = (close_num[j, k] - muhat[j, k]) / sigmahat[j, k]
+            # Calculate both positive and negative enrichment p values
+            p[j, k, 0] = (1 + (np.sum(tmp >= close_num[j, k]))) / (bootstrap_num + 1)
+            p[j, k, 1] = (1 + (np.sum(tmp <= close_num[j, k]))) / (bootstrap_num + 1)
+
+    # Get fdh_br adjusted p values
+    p_summary = np.zeros_like(p[:, :, 0])
+    for j in range(0, marker_num):
+        for k in range(0, marker_num):
+            # Use negative enrichment p values if the z score is negative, and vice versa
+            if z[j, k] > 0:
+                p_summary[j, k] = p[j, k, 0]
+            else:
+                p_summary[j, k] = p[j, k, 1]
+    (h, adj_p, aS, aB) = statsmodels.stats.multitest.multipletests(
+        p_summary, alpha=.05)
+
+    return z, muhat, sigmahat, p, h, adj_p
+
+
+def calculate_channel_spatial_enrichment(dist_matrix, marker_thresholds, all_patient_data,
+                                         excluded_colnames=None, points=None,
+                                         patient_idx=30, dist_lim=100, cell_label_idx=24, bootstrap_num=100):
+    """Spatial enrichment analysis to find significant interactions between cells expressing different markers.
```

Comment:
Good. Because this is more than one line, the best practice is to have a single line that describes at a very high level what the function does, then any additional lines separated below that give more specifics